src/backend/distributed/replication/multi_logical_replication.c

```diff
@@ -1232,177 +1523,167 @@ GetQueryResultStringList(MultiConnection *connection, char *query)
+DropAllPublications(MultiConnection *connection, LogicalRepType type)
 {
 	char *query = psprintf(
 		"SELECT pubname FROM pg_publication "
 		"WHERE pubname LIKE %s || '%%'",
-		quote_literal_cstr(SHARD_MOVE_PUBLICATION_PREFIX));
+		quote_literal_cstr(publicationPrefix[type]));
 	List *publicationNameList = GetQueryResultStringList(connection, query);
 	char *publicationName;
 	foreach_ptr(publicationName, publicationNameList)
 	{
-		DropShardPublication(connection, publicationName);
+		DropPublication(connection, publicationName);
 	}
 }
 
 
 /*
- * DropShardMoveSubscriptions drops subscriptions from the subscriber node that
- * are used to move shards for the given table owners. Note that, it drops the
- * replication slots on the publisher node if it can drop the slots as well
- * with the DROP SUBSCRIPTION command. Otherwise, only the subscriptions will
- * be deleted with DROP SUBSCRIPTION via the connection. In the latter case,
- * replication slots will be dropped while cleaning the publisher node when
- * calling DropShardMovePublications.
+ * DropSubscriptions drops all the subscriptions in the subscriptionInfoList
+ * from the subscriber node. It also drops the temporary users that are used as
+ * owners for of the subscription. This doesn't drop the replication slots on
+ * the publisher, these should be dropped using DropReplicationSlots.
  */
-static void
-DropShardMoveSubscriptions(MultiConnection *connection, Bitmapset *tableOwnerIds)
+void
+DropSubscriptions(List *subscriptionInfoList)
 {
-	int ownerId = -1;
-	while ((ownerId = bms_next_member(tableOwnerIds, ownerId)) >= 0)
+	SubscriptionInfo *subscription = NULL;
+	foreach_ptr(subscription, subscriptionInfoList)
 	{
-		DropShardSubscription(connection, ShardSubscriptionName(ownerId,
-																SHARD_MOVE_SUBSCRIPTION_PREFIX));
-		DropShardUser(connection, ShardSubscriptionRole(ownerId,
-														SHARD_MOVE_SUBSCRIPTION_ROLE_PREFIX));
+		DropSubscription(subscription->targetConnection,
+						 subscription->name);
+		DropUser(subscription->targetConnection, subscription->temporaryOwnerName);
 	}
 }
 
 
 /*
- * DropShardSubscription drops subscription with the given name on the
- * subscriber node. Note that, it also drops the replication slot on the
- * publisher node if it can drop the slot as well with the DROP SUBSCRIPTION
- * command. Otherwise, only the subscription will be deleted with DROP
- * SUBSCRIPTION via the connection.
+ * DropSubscription drops subscription with the given name on the subscriber
+ * node if it exists. Note that this doesn't drop the replication slot on the
+ * publisher node. The reason is that sometimes this is not possible, usually
+ * due to the node with the replication slot being down. So instead of directly
+ * dropping the subscription, it is first disconnected from the replication
+ * slot before dropping it. The replication slot itself should be dropped using
+ * DropReplicationSlot on the source connection.
```

Comment:
Also the reason to disable the subscriber, reset slot name and then drop is to avoid deadlock when the subscription is local. This is the case with local shard splits.

Reply:
Thanks for that insight, updated the the comment accordingly.